### ISSUE TITLE
When in any type of list, automatically add new list items

### ISSUE THREAD

**Issue by MatanRubin** [fabiospampinato/notable#88]
When writing down a list, it would be useful to automatically create the next list item when the user presses Enter. Press Enter one more time to delete the list item.

This should work in bulleted lists, numbered lists, and task lists.
This mimics list behavior in Office products, Gmail, and most text editing software.

**Comment by fabiospampinato** [comment]
I'm not sure I like this, if you're currently typing the last item on the list this can get annoying quickly.

I think maybe this could be implemented but configurable and disabled by default.

If anyone would like to take a shot at this I'll be happy to consider a PR for it.

**Comment by sebastiendecraene** [comment]
@fabiospampinato where should I put the setting for this? Is this in the `Settings.ts` file? Also what would you like to call this setting?

**Comment by fabiospampinato** [comment]
@sebastiendecraene it depends on the implementation, how are you implementing this? ideally I would add this as an option to CodeMirror and then expose a why of override any CodeMirror setting.

**Comment by sebastiendecraene** [comment]
@fabiospampinato 
https://github.com/fabiospampinato/notable/blob/5c1b74b7d36ba639dab81b5177bac9f45dc76559/src/renderer/components/main/mainbar/code/index.tsx#L44-L49
I added `'Enter': Utils.onEnterPressed` at the end of this. And then in the Utils I used `cm.execCommand("newlineAndIndent");` to make sure the default action is happening. Then I was planning on checking if the feature is enabled.

**Comment by MatanRubin** [comment]
> I'm not sure I like this, if you're currently typing the last item on the list this can get annoying quickly.

Well, most lists have multiple items which are not the last item in the list, and only one item which is the last item of the list, that's why gmail etc. optimize for the common case and make it easy to insert a new item to the list. To end the list you simply press Enter twice, which isn't a real hassle.

**Comment by sebastiendecraene** [comment]
> > I'm not sure I like this, if you're currently typing the last item on the list this can get annoying quickly.
> 
> Well, most lists have multiple items which are not the last item in the list, and only one item which is the last item of the list, that's why gmail etc. optimize for the common case and make it easy to insert a new item to the list. To end the list you simply press Enter twice, which isn't a real hassle.

That's actually not a bad idea.

**Comment by fabiospampinato** [comment]
@sebastiendecraene isn't there an already-made extension that does this for CodeMirror? the less code to maintain the better.

IMHO it should work like this:

- Upon `Enter` the indentation is preserved, along with the bullet character is preserved
- If it's an ordered list the number should be incremented
- If `Enter` is pressed again we remove the added indentation/characters

Regarding the setting I'm not sure what the best option is, let's think about that later.

**Comment by MatanRubin** [comment]
> isn't there an already-made extension that does this for CodeMirror?
Haven't checked myself, but given how common this behavior is everywhere, I suspect there is one.

> * If `Enter` is pressed again we remove the added indentation/characters
The standard behavior elsewhere is that each additional Enter decreases the level of indentation, and only the last one removes the bullet completely. This helps you easily navigate the indentation levels of nested lists. You can play with Gmail's lists to get the exact semantics.

**Comment by fabiospampinato** [comment]
> The standard behavior elsewhere is that each additional Enter decreases the level of indentation, and only the last one removes the bullet completely. This helps you easily navigate the indentation levels of nested lists. You can play with Gmail's lists to get the exact semantics.

Makes sense. Yeah as rule of thumb we should pretty much implement it the same way that gmail does it.

**Comment by sebastiendecraene** [comment]
I'll try my best :)

**Comment by TwisterMc** [comment]
I'd love this feature too. When taking notes in meetings, I use bullet lists a lot to take notes quicker.

**Comment by sebastiendecraene** [comment]
I've written some code for it but it's far from perfect, but the basic stuff is there. I don't have time right now to complete it because i'm in the middle of my exams: https://github.com/sebastiendecraene/notable/tree/9be9cb1f276842417ebc6044b8111c6afc2978bd/src/renderer/components/main/mainbar/code This is where I've written the basic implementation

**Comment by MatanRubin** [comment]
@sebastiendecraene, I've added a commit to your commits and open a PR, see https://github.com/fabiospampinato/notable/pull/198

**Comment by karbecker** [comment]
+1 from me... Typora does this nicely and especially for task lists it's really comfortable IMO:

![feature](https://user-images.githubusercontent.com/8436596/54599103-3ccae800-4a3a-11e9-913b-191bb215e735.gif)

**Comment by fabiospampinato** [comment]
We recently switched to Monaco #203, if somebody would like to work an a PR for this I'd be happy to consider it 👍

**Comment by awmottaz** [comment]
Not really sure about what's involved, but would it be possible to port over the [Markdown All-in-one VSCode extension](https://github.com/yzhang-gh/vscode-markdown) here? If so, that would solve this issue and add lots of other nice functionality.

**Comment by fabiospampinato** [comment]
@awmottaz Not currently. Also I think most of the features provided by that extension should come built-in here. But I'll definitely look into what it would take to support VSCode extensions here (we are using the same editor after all) as that would be interesting.

**Comment by awmottaz** [comment]
> Also I think most of the features provided by that extension should come built-in here.

I agree. I wasn't considering supporting VSCode extensions (I also think that sounds interesting), but rather reusing the code from that repo to reproduce the functionality natively.

When I get some time, I might give it a go

**Comment by mHatsune** [comment]
![beta](https://user-images.githubusercontent.com/63988538/86993455-068d1b00-c1df-11ea-8bb5-fb51e84012c7.gif)

**Comment by exislow** [comment]
This is SUCH a **mandatory** feature. I would highly appreaciate this.

Please also consider to intend the bullet point / enumeration etc. if you press tab. It should be intended and the type of bullet point enumeration should change like this, e.g.

```
*
    +
         -

1.
    a.
...
```

Are you planning to release it with the next release?

**Comment by fabiospampinato** [comment]
> Please also consider to intend the bullet point / enumeration etc. if you press tab. It should be intended and the type of bullet point enumeration should change like this, e.g.

I'm not sure how many people would want that 🤔 are you currently using another app that implements it? Also "a." is just not a valid way to make ordered lists in Markdown.

> Are you planning to release it with the next release?

No.

**Comment by DamienHodsall** [comment]
Is this still in the works? A feature like this could be really really helpful. I think having it optional and off by default would be perfect for most users.

**Comment by MMK21Hub** [comment]
From the Discord server:
> Not super soon, but it's something I want to get to eventually
([Message sent by @ fa bios pampinato#6061 on 12 September 2022](https://discord.com/channels/715934079559663646/720653307747500092/1018826080741773342))

**Comment by teleksterling** [comment]
I'd propose that this continuation behaviour (once implemented) also include task list formatting.
